bond_app/bond.py

```diff
@@ -60,23 +60,61 @@ def exchange_authz_code(self, authz_code, redirect_uri, sam_user_id):
                                       jwt_token.username, self.provider_name)
         return jwt_token.issued_at, jwt_token.username
 
-    def generate_access_token(self, sam_user_id):
+
+    def generate_access_token(self, sam_user_id, refresh_token=None):
         """
-        Given a user, lookup their refresh token and use it to generate a new refresh token from their OAuth
+        Given a user, lookup their refresh token (if not provided) and use it to generate a new refresh token (<-should this be "access token"?) from their OAuth
         provider.  If a refresh token cannot be found for the sam_user_id provided, a NotFound will be raised.
         :param sam_user_id: Id stored in Sam for user who initiated request
         :return: Two values: An Access Token string, datetime when that token expires
         """
-        refresh_token = self.refresh_token_store.lookup(sam_user_id, self.provider_name)
+        refresh_token = refresh_token or self.refresh_token_store.lookup(sam_user_id, self.provider_name)
         if refresh_token is not None:
             token_response = self.oauth_adapter.refresh_access_token(refresh_token.token)
             expires_at = datetime.fromtimestamp(token_response.get(FenceKeys.EXPIRES_AT))
+
             return token_response.get("access_token"), expires_at
         else:
             raise exceptions.NotFound(
                 "Could not find refresh token for sam_user_id: {} provider_name: {}\nConsider relinking your account to Bond.".format(
                     sam_user_id, self.provider_name))
 
+
+    def get_access_token(self, sam_user_id, refresh_token=None, refresh_threshold: int = 600):
+        refresh_token = refresh_token or self.refresh_token_store.lookup(sam_user_id, self.provider_name)
+        if refresh_token is not None:
+            cached_access_entry = self.fence_tvm.cache_api.get_entry(namespace="AccessTokens", key=sam_user_id)
```

Comment:
It seemed reasonable at the time, but now I'm not sure that the `FenceTokenVendingMachine` is the right place to access this cache API. Let me know what you think...

Reply:
Honest question and not (yet?) veiled suggestion: Did you consider adding a structure like the `[Refresh]TokenStore` to hold access tokens? 

I'm having trouble tracking the various types of tokens and places we can stick data, but I'm not sure it feels natural to put them here. It looks like the cache api on `FenceTokenVendingMachine` wasn't actually used before this change, which makes me suspicious.

Reply:
I didn't consider that, but it makes sense. The `TokenStore` class seems to have a lot in common with `DatastoreCacheApi`, almost like it evolved in parallel, or one represents a newer vision of how tokens will be cached in the future.

> It looks like the cache api on FenceTokenVendingMachine wasn't actually used before this change, which makes me suspicious.

100% agree with this.

It's probably simplest to create a parallel `AccessTokenStore` class, but we may want someone more familiar with a higher level view of this repository to chime in about the direction we should go.

Reply:
I would not use `self.fence_tvm.cache_api` but rather pass in a `cache_api` instance to this class via constructor

Reply:
thanks @dvoet, that sounds good to me.

I am a bit curious about `TokenStore` vs `DatastoreCacheApi`. But if that's out-of-scope, it's all good with me.